### PR TITLE
[JBPM-9195] Case traceability from a Task Instance

### PR DESCRIPTION
**Pull request by fjtirado**
CorrelationKey and ProcessType added to UserTaskDescription.

Depends on [droolsjbpm-knowledge:447](https://github.com/kiegroup/droolsjbpm-knowledge/pull/447) and merge together with https://github.com/kiegroup/droolsjbpm-integration/pull/2136

Existing APIs(getTaskById, getTaskByWorkItem and getTaskByProcess) are
reused, since I believe is preferable to assume the performance cost of
an extra join rather than having to maintain a new API for every
attribute combination (specially when there is already a join because of
the introduction of subject attribute)

### CODE REVIEW

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by elguardian** [comment]
Jenkins retest this please.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 1]
```diff
@@ -1039,19 +1040,23 @@ public void testGetTaskById() {
...
-		assertNotNull(userTask);
-
-		assertEquals(nodeInstanceDesc.getSlaCompliance(), userTask.getSlaCompliance());
-		assertEquals(nodeInstanceDesc.getSlaDueDate(), userTask.getSlaDueDate());
+    	userTask = runtimeDataService.getTaskById(taskId, true);
+    	assertNotNull(userTask);
+    	assertEquals(nodeInstanceDesc.getSlaCompliance(), userTask.getSlaCompliance());
+    	assertEquals(nodeInstanceDesc.getSlaDueDate(), userTask.getSlaDueDate());
+    	
+    	assertNotNull(userTask.getCorrelationKey());
```
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1051]: Not enought check also the value of correlation key. as a rule of thumb the correlation key by default is the process instance id to string but you can start a process with a custom correlation key. (do that)

File: jbpm-services/jbpm-services-cdi/src/test/java/org/jbpm/services/cdi/test/RuntimeDataServiceTest.java [hunk 2]
```diff
@@ -266,7 +267,7 @@ public void testGetTasksAssignedAsPotentialOwnerGroupFromCallback() {
         List<TaskSummary> tasks = runtimeDataService.getTasksAssignedAsPotentialOwnerByStatus("katy", statuses, new QueryFilter());
         assertNotNull(tasks);
         assertEquals(1, tasks.size());
-        
+        assertEquals(WorkflowProcess.PROCESS_TYPE, tasks.get(0).getProcessType().intValue());
```
Comment (elguardian) [inline jbpm-services/jbpm-services-cdi/src/test/java/org/jbpm/services/cdi/test/RuntimeDataServiceTest.java:270]: what about the correlation key ?

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/resources/META-INF/CaseMgmtorm.xml [hunk 3]
```diff
@@ -444,7 +444,7 @@
                 t.taskData.processInstanceId,
                 t.taskData.parentId,
                 t.taskData.deploymentId,
-                t.taskData.skipable              )
+                t.taskData.skipable, plog.correlationKey, plog.processType)
```
Comment (gmunozfe) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/resources/META-INF/CaseMgmtorm.xml:447]: It should be "log" instead of "plog" as alias
```suggestion
                t.taskData.skipable, log.correlationKey, log.processType)
```

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/resources/META-INF/CaseMgmtorm.xml [hunk 4]
```diff
@@ -478,7 +478,7 @@
...
```
Comment (gmunozfe) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/resources/META-INF/CaseMgmtorm.xml:481]: Same here: It should be "log" instead of "plog" as alias
```suggestion
                t.taskData.skipable, log.correlationKey, log.processType)
```

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
@kiegroup/gatekeepers please merge

**Comment by mareknovotny** [comment]
@fjtirado there is no approval from reviewers, we need to wait for that first

**Comment by fjtirado** [comment]
I know, Enrique told me to do that. I asked him to re-review.

Enviado desde Mailspring (https://link.getmailspring.com/link/A10EEF90-F53A-4664-8F10-8AAAE645234A@getmailspring.com/0?redirect=https%3A%2F%2Fgetmailspring.com%2F&recipient=cmVwbHkrQVBSWFlQVkhVTVQ1NzZMUktSVlZSWTU1Qkw3SFJFVkJOSEhDTUs2UEhFQHJlcGx5LmdpdGh1Yi5jb20%3D), la mejor aplicación de correo electrónico gratuita para trabajar

**Comment by mareknovotny** [comment]
Jenkins execute compile downstream

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by mareknovotny** [comment]
there are compiler errors in PR check, please fix them
```
 [2020-07-07T17:00:38.931Z] [ERROR] COMPILATION ERROR : 
[2020-07-07T17:00:38.931Z] [INFO] -------------------------------------------------------------
[2020-07-07T17:00:38.931Z] [ERROR] /home/jenkins/workspace/KIE/master/pullrequest/new-PR-jbpm-master.pullrequests/jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/impl/model/TaskEventImpl.java:[155,3] method does not override or implement a method from a supertype
[2020-07-07T17:00:38.932Z] [ERROR] /home/jenkins/workspace/KIE/master/pullrequest/new-PR-jbpm-master.pullrequests/jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/impl/model/TaskEventImpl.java:[160,3] method does not override or implement a method from a supertype
[2020-07-07T17:00:38.932Z] [ERROR] /home/jenkins/workspace/KIE/master/pullrequest/new-PR-jbpm-master.pullrequests/jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/impl/model/TaskEventImpl.java:[165,3] method does not override or implement a method from a supertype
[2020-07-07T17:00:38.932Z] [ERROR] /home/jenkins/workspace/KIE/master/pullrequest/new-PR-jbpm-master.pullrequests/jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/impl/model/TaskEventImpl.java:[170,3] method does not override or implement a method from a supertype
[2020-07-07T17:00:38.932Z] [INFO] 4 errors 
```

**Comment by fjtirado** [comment]
> compiler errors need to be fixed

There is a dependency betwen jbpm and  knowledge repository. TaskEvent class was missing from the commit on Knowledge repostiroy  Now that knowledge have been properly updated, compilation errors should dissapear from jbpm

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
jenkins retest this

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java [hunk 5]
```diff
@@ -1131,11 +1131,15 @@ private void addVariableNameAndVariableValueParameters(TaskSummaryQueryBuilder q
...
-        if(taskEvents == null || taskEvents.isEmpty()) {
-            
-            UserTaskInstanceDesc task = getTaskById(taskId);
-            if (task == null) {
-                throw new TaskNotFoundException( MessageFormat.format(TASK_NOT_FOUND, taskId) );
+        
+        UserTaskInstanceDesc task = getTaskById(taskId);
+        if (task == null) {
+            throw new TaskNotFoundException( MessageFormat.format(TASK_NOT_FOUND, taskId) );
+        }
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java:1138]: This check can now be made in the beginning of the method as it would eliminate possibly unnecessary querying of all task events in case the task doesn't even exist.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 6]
```diff
@@ -1016,12 +1017,15 @@ public void testGetTaskByWorkItemId() {
     	assertNotNull(userTask);
     	assertEquals(processInstanceId, userTask.getProcessInstanceId());
     	assertEquals("Write a Document", userTask.getName());
+        assertTask(userTask, processInstanceId);
 
     }
 
     @Test
     public void testGetTaskById() {
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1025]: Can you create a new test for this as this is modifying the old one? It uses a different startProcess method.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 7]
```diff
@@ -1118,26 +1127,20 @@ public void testGetTaskAssignedAsBusinessAdminPaging() {
...
-        }
+        startProcesses(10, "org.jbpm.writedocument");
         
         List<Status> statuses = new ArrayList<Status>();
         statuses.add(Status.Ready);
         statuses.add(Status.Reserved);
 
         List<TaskSummary> tasks = runtimeDataService.getTasksAssignedAsBusinessAdministratorByStatus("Administrator", statuses, new QueryFilter(0, 5));
-        assertNotNull(tasks);
-        assertEquals(5, tasks.size());
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1133]: Why was this removed? It should still hold I think.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 8]
```diff
@@ -1210,22 +1213,15 @@ public void testGetTasksAssignedAsPotentialOwnerByStatusPagingAndFiltering() {
...
         List<Status> statuses = new ArrayList<Status>();
         statuses.add(Status.Ready);
         statuses.add(Status.Reserved);
 
         QueryFilter ctx = new QueryFilter(0, 5, "Status", true);
         
         List<TaskSummary> tasks = runtimeDataService.getTasksAssignedAsPotentialOwnerByStatus("salaboy", statuses, ctx);
-        assertNotNull(tasks);
-        assertEquals(5, tasks.size());
-
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1228]: Same here.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 9]
```diff
@@ -1329,12 +1325,15 @@ public void testGetTaskEvents() {
...
 
     	List<TaskEvent> auditTasks = runtimeDataService.getTaskEvents(userTask.getTaskId(), new QueryFilter());
     	assertNotNull(auditTasks);
-        assertEquals(2, auditTasks.size());
-    	assertEquals(TaskEvent.TaskEventType.ADDED, auditTasks.get(0).getType());
+    	assertEquals(2, auditTasks.size());
+    	TaskEvent taskEvent = auditTasks.get(0);
+    	assertEquals(TaskEvent.TaskEventType.ADDED, taskEvent.getType());
+    	assertNotNull(taskEvent.getCorrelationKey());
+    	assertNotNull(taskEvent.getProcessType());
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1336]: Maybe assert the process type as well.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java [hunk 10]
```diff
@@ -110,4 +115,33 @@ protected KieFileSystem createKieFileSystemWithKProject(KieServices ks) {
         return kfs;
     }
 
+    protected void assertTask(UserTaskInstanceDesc userTask, long processInstanceId) {
+        assertTask(userTask, Long.toString(processInstanceId));
+    }
+
+    protected void assertTask(UserTaskInstanceDesc userTask, String correlationKey) {
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java:122]: Maybe make the name more descriptive - e.g. what it's asserting on.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 11]
```diff
@@ -1239,7 +1235,7 @@ public void testGetTasksAssignedAsPotentialOwnerSortedByAlias() {
     @Test
     public void testTasksByStatusByProcessInstanceIdPagingAndFiltering() {
 
-    	Long pid = processService.startProcess(deploymentUnit.getIdentifier(), "org.jbpm.writedocument");
+    	processInstanceId = processService.startProcess(deploymentUnit.getIdentifier(), "org.jbpm.writedocument");
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1238]: I would rather use a variable which belongs to the test itself as it's less error-prone to tests influencing each other.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java [hunk 12]
```diff
@@ -110,4 +115,33 @@ protected KieFileSystem createKieFileSystemWithKProject(KieServices ks) {
         return kfs;
     }
 
+    protected void assertTask(UserTaskInstanceDesc userTask, long processInstanceId) {
+        assertTask(userTask, Long.toString(processInstanceId));
+    }
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java:120]: Why do we need this intermediate step and not do the conversion inside the assert method below? As this method doesn't do anything more.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java (hunk 5 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java:1138]: yes, good point

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 7 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1133]: There was an already existing issue with this test. 
Basically Task table is not removed between test, which leads to inconsistencies. 
Since it is apparently not related with these changes, Gonzalo will investigate this as a separate task.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 8 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1228]: There was an already existing issue with this test. 
Basically Task table is not removed between test, which leads to inconsistencies. 
Since it is apparently not related with these changes, Gonzalo will investigate this as a separate task.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 9 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1336]: processType here can be either PROCESS or CASE, so what is tested is that TaskEvent is filled as expected with a not null value

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java (hunk 10 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java:122]: Done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 11 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1238]: I agree with the principle, but in practise global variable approach is used for all other test and allow defining commom method in parent class (changing that at this stage is not providig any benefit). This line was modified  to align this test with the other ones.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java (hunk 12 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java:120]: This method is a bridge between processInstanceId and correlationKey. In most test, correlationKey is equal to processInstanceId. This is doing the briding. If this method is removed Long.toString will have to be invoked in a lot of test.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 6 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1025]: I think once the name of asserTask method has been changed to assertCorrelationAndProcess, it is now clear the pupose of adding this line here. To test that for getTaskByWorkItemId, processType and correlationKey are included in the response

Comment (MarianMacik) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1025]: I mean the test which I commented `testGetTaskById()` :) As you changed the behaviour of the test - not that good for regression testing. (For example you use now a different startProcess method) So I would say that a separate test would be better.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 9 above)
Comment (MarianMacik) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1336]: But since you started process in this test you can expect PROCESS here, so the test would have even better coverage.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java (hunk 12 above)
Comment (MarianMacik) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/test/util/AbstractKieServicesBaseTest.java:120]: I see now, thanks.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 11 above)
Comment (MarianMacik) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java:1238]: OK, makes sense.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 6 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: Ok, understood ;)
New test added.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 9 above)
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: you are right, changed

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/mapper/UserTaskInstanceQueryMapper.java [hunk 13]
```diff
@@ -80,7 +80,11 @@ protected UserTaskInstanceDesc buildInstance(DataSet dataSetResult, int index) {
                 getColumnDateValue(dataSetResult, COLUMN_DUEDATE, index),//dueDate
                 getColumnLongValue(dataSetResult, COLUMN_WORKITEMID, index), //workItemId
                 getColumnStringValue(dataSetResult, COLUMN_FORM_NAME, index),
-                getColumnStringValue(dataSetResult, COLUMN_SUBJECT, index)
+                getColumnStringValue(dataSetResult, COLUMN_SUBJECT, index),
+                getColumnStringValue(dataSetResult, COLUMN_CORRELATIONKEY, index), 
+                getColumnIntValue(dataSetResult, COLUMN_PROCESSTYPE, index),
+                getColumnDateValue(dataSetResult, COLUMN_SLA_DUE_DATE, index), 
+                getColumnIntValue(dataSetResult, COLUMN_SLA_COMPLIANCE, index)
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/mapper/UserTaskInstanceQueryMapper.java:87]: SLA-related columns are used for node(task)-related SLA (from NodeInstanceLog), not for process SLA (from ProcessInstanceLog). These fields are used for the `RuntimeDataServiceImpl#getTaskById(Long taskId, boolean withSLA)` method where you can specify if you want just the task or also its SLA data. So by default it's fine that the mapper doesn't have this data.

At first, the join with NodeInstanceLog was implemented in https://github.com/kiegroup/jbpm/commit/3c4676575601921353e69b41e26c2fc61fd1cedc#diff-031cb54fb8b11dac505ec36874a81a26 but then we found that it negatively impacts performance so we decided to create a separate method for that and reuse the same `UserTaskInstanceDesc` object for that. That's why SLA information is not in the mapper by default.

Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/mapper/UserTaskInstanceQueryMapper.java]: Ok, adding two separate set of contanst for these two columns on task table, in case they are retrieved in future

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
@kiegroup/gatekeepers could you please merge?

**Comment by MarianMacik** [comment]
@mareknovotny Can you approve as well? Thanks.

**Comment by mareknovotny** [comment]
@MarianMacik well, this is probably from me just formal as I am not completely knowledgeable person.
But one thing I would like to see here is FDB run
